src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -8,46 +8,48 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Serialization;
 using Microsoft.CodeAnalysis.SolutionCrawler;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     /// <summary>
-    /// primary workspace for remote host. no one except solution service can update this workspace
+    /// Workspace created by the remote host that mirrors the corresponding client workspace.
     /// </summary>
-    internal class RemoteWorkspace : Workspace
+    internal sealed class RemoteWorkspace : Workspace
     {
         private readonly ISolutionCrawlerRegistrationService? _registrationService;
 
         // guard to make sure host API doesn't run concurrently
         private readonly object _gate = new object();
 
+        private readonly SemaphoreSlim _solutionSynchronizationGate = new SemaphoreSlim(initialCount: 1);
+
+        // the last solution for the the primary branch fetched from the client
+        private volatile Tuple<Checksum, Solution>? _primaryBranchSolutionWithChecksum;
+
+        // the last solution requested by a service:
+        private volatile Tuple<Checksum, Solution>? _lastRequestedSolutionWithChecksum;
+
         // this is used to make sure we never move remote workspace backward.
         // this version is the WorkspaceVersion of primary solution in client (VS) we are
         // currently caching
         private int _currentRemoteWorkspaceVersion = -1;
 
-        public RemoteWorkspace()
-            : this(applyStartupOptions: true)
-        {
-        }
-
         // internal for testing purposes.
-        internal RemoteWorkspace(bool applyStartupOptions)
-            : base(RoslynServices.HostServices, workspaceKind: WorkspaceKind.RemoteWorkspace)
+        internal RemoteWorkspace(HostServices hostServices, string? workspaceKind)
+            : base(hostServices, workspaceKind)
         {
             var exportProvider = (IMefHostExportProvider)Services.HostServices;
-            var primaryWorkspace = exportProvider.GetExports<PrimaryWorkspace>().Single().Value;
-            primaryWorkspace.Register(this);
-
             RegisterDocumentOptionProviders(exportProvider.GetExports<IDocumentOptionsProviderFactory, OrderableMetadata>());
 
-            if (applyStartupOptions)
-                SetOptions(Options.WithChangedOption(CacheOptions.RecoverableTreeLengthThreshold, 0));
+            SetOptions(Options.WithChangedOption(CacheOptions.RecoverableTreeLengthThreshold, 0));
```

Comment:
really curious what the point of this option is at all.

Reply:
I do not know.